### ISSUE TITLE
Json Matchers Supporting Custom `Json`

### ISSUE THREAD

**Issue by solonovamax** [kotest/kotest#4601]
It would be nice if there were additional json matchers allowing a kotlinx.serialization `Json` to be passed to them.

For example:
```kotlin
fun beValidJson(parser: Json) = object : Matcher<String?> {
   override fun test(value: String?): MatcherResult {
      return try {
         value?.let(parser::parseToJsonElement)
         MatcherResult(
            true,
            { "expected: actual json to be valid json: $value" },
            { "expected: actual json to be invalid json: $value" }
         )
      } catch (ex: Exception) {
         MatcherResult(
            false,
            { "expected: actual json to be valid json: $value" },
            { "expected: actual json to be invalid json: $value" }
         )
      }
   }
}

fun String.shouldBeValidJson(parser: Json): String {
   this should beValidJson(parser)
   return this
}

fun String.shouldNotBeValidJson(parser: Json): String {
   this shouldNot beValidJson()
   return this
}
```

An example usecase of this is if I want to validate that the specified json is valid when `isLenient` is set to `true`.
Another usecase would be providing a custom `serializersModule`.

**Comment by sksamuel** [comment]
If it's not valid json, then you wouldn't be able to make a Json object from it ?

**Comment by solonovamax** [comment]
> If it's not valid json, then you wouldn't be able to make a Json object from it ?

it's not valid json according to the spec, however it can still be parsed according to kotlinx.serialization

eg.
```json5
{
    // this is json with a comment
    "foo": "bar"
}
```

this can be parsed by kotlinx.serialization if you enable comment support

**Comment by sksamuel** [comment]
So the matcher would be the json is invalid according to the json spec, but valid according to kotlinx ?

**Comment by solonovamax** [comment]
> So the matcher would be the json is invalid according to the json spec, but valid according to kotlinx ?

yes, correct

**Comment by sksamuel** [comment]
I don't think we should add a matcher that checks if kotlinx.serialization is following the spec. Should file a bug over there?

**Comment by solonovamax** [comment]
> I don't think we should add a matcher that checks if kotlinx.serialization is following the spec. Should file a bug over there?

I don't want to check if kotlinx.serialization is following the spec, because as far as I know it is (unless you explicitly instruct it not to)

what I want to check is "if I tell kotlinx.serialization to not follow the spec, can my not-following-spec still json be parsed properly within the bounds of what kotlinx.serialization can deserialize"

for example, I would expect that this fails
```json
{
    "foo
```

but this passes
```json5
{
    // this is a comment
    "foo": "bar",
}
```

when parsed with the following `Json` parser options:
```kotlin
Json {
    isLenient = true
    allowComments = true
    allowTrailingComma = true
}
```

however, there is no way to specify to kotest that I want to parse the string using that `Json` parser.

---

or, let's say I wanted to check if I can properly deserialize the following json:
```json
{
    "name": "Kotlin",
    "stableReleaseDate": 1455494400000
}
```
for this class:
```kotlin
@Serializable          
class ProgrammingLanguage(
    val name: String,
    @Contextual 
    val stableReleaseDate: Date
)
```

as you can see, the `Date` field annotated with `@Contextual`. this is because `Date` is not serializeable, and it allows for the serializer to be looked up at runtime rather than included in the generated serializer code.
But, it would require configuring a custom `Json` parser:
```kotlin
val parser = Json {
    serializersModule = SerializersModule {
        contextual(DateAsLongSerializer)
    }
}
```
where `DateAsLongSerializer` is a serializer for `Date`, which is able to serialize/deserialize it as a `Long`.

but if I were attempt to validate this json with
```kotlin
val result: String = TODO()
result should beJsonType(ProgrammingLanguage::class)
```

this test would fail, even though it is valid json and in the context of the program it is deserializable, because a custom `SerializersModule` is always used.

**Comment by Kantis** [comment]
@solonovamax do you want to work on adding this?

**Comment by PreAgile** [comment]
Hi @solonovamax, are you still planning to work on this? @Kantis mentioned it back in Jan 2025 but I don't see a
  PR yet.

  If not, I'd be happy to give it a go. My plan would be to add a parser: Json = pretty default parameter to the
  matcher functions so existing callers aren't affected. Let me know if that approach sounds right.

**Comment by sksamuel** [comment]
Go for it @PreAgile
